proposals/20230530-driver-kernel-testing-framework.md

```diff
@@ -0,0 +1,453 @@
+
+
+## Challenges and Considerations
+
+Ensuring compatibility between different kernel versions, distributions, architectures, and compiler versions is complex due to frequent updates and changes in the Linux kernel. 
+
+The curse of dimensionality and combinatorial explosion present challenges in kernel testing, as the testing complexity exponentially increases with the number of dimensions. Testing every possible combination exhaustively becomes impractical. However, we have an advantage in that most distributions only change kernels from one LTS version to the next LTS, while frequently publishing new builds (kernel releases). In addition, compatibility issues tend to be tied to major and minor kernel version changes rather than individual kernel releases. As a result, the realistic grid of potential kernels becomes more manageable and tractable.
+
+Performance and impact can vary based on specific workloads and fine-grained kernel settings, which are difficult to predict for each adopter's unique use case. Therefore, a strategic approach is crucial to balance coverage and resources in testing.
+
+Furthermore, considering real-world workloads and settings is essential for comprehensive testing and accurate performance evaluation, serving as the guiding principle throughout the process.
+
+
+## Proposal
+
+The "Functionality" tests of the CI-powered test framework are targeted for Falco release 0.36.
+
+- Objective: Ensure Falco drivers run and work across different kernel versions and distributions.
+- Actions: Implement a comprehensive testing process and kernel test grid to validate driver compatibility, stability, and functionality across various kernel environments.
+- Expected Outcome: Increased confidence in the reliability and stability of Falco drivers.
+
+### Test Category 1
+
+Ensuring that the kernel driver successfully compiles for the agreed-upon kernel test grid. The optimal compiler version and build container are carefully selected to align with the advancements in the Linux source tree and its related dependencies. This selection takes into account various factors, including the specific kernel version, driver type, and the distribution being used.
+
+
+### Test Category 2
+
+Verifying that the kernel driver can load, run, and capture events without errors. This is determined through [scap-open](https://github.com/falcosecurity/libs/tree/master/userspace/libscap/examples/01-open) and unit tests conducted in virtual machine (VM) environments. In essence, when we mention that the "driver loads and runs", it implies that the scap-open counter for captured events during a test run is positive and that the [drivers_test](https://github.com/falcosecurity/libs/tree/master/test/drivers) unit tests pass. The latter tests not only load the driver live but also simulate syscall events and verify that the expected information is extracted from the kernel tracepoint and retrieved by the libscap driver type-specific engine in userspace.
```

Comment:
> run is positive and that the drivers_test unit tests

Just one minor note about this. We know that on older kernel versions `~<5.0` driver_tests could fail due to conflicting events captured. Let me explain better, is possible that while we search for a `close` event spawned by our test we face another `clone` event generated by the system causing a test failure (as far as i saw this is very frequent when we use `tracepoints` instead of `raw_traceoints` in the old probe, so in kernels between `4.14` and `4.17`) ... so probably we won't be able to rely on the drivers_test on all machines as a source of truth. We will try to fix driver_tests to avoid these failures but in the meanwhile, I would consider scap-open as the only source of truth and use the drivers_tests as a best-effort check. WDYT?

Reply:
Thanks @Andreagit97 I agree, `driver_tests` should be best effort at first and we can stabilize it subsequently.